### PR TITLE
feature/COR-1915_visual-improvements-to-x-axis

### PR DESCRIPTION
**Pull request by VWSCoronaDashboard30**
Refactored logic for timeframe behavior of single and date range ticks

## Summary
* 
*
*
 
#### Before
<details>
<summary>Toggle before screenshots</summary>
 
</details>
 
#### After
<details>
<summary>Toggle after screenshots</summary>
 
</details>

### CODE REVIEW

**Comment by ben-van-eekelen** [comment]
@VWSCoronaDashboard30 can you add some screenshots of before and after?

File: packages/app/src/components/age-demographic/age-demographic-chart.tsx
Comment (ben-van-eekelen) [inline packages/app/src/components/age-demographic/age-demographic-chart.tsx:1]: Did you re-order the imports too?

File: packages/app/src/components/time-series-chart/components/axes/components/axes.tsx
Comment (ben-van-eekelen) [inline packages/app/src/components/time-series-chart/components/axes/components/axes.tsx:1]: Why did you move this file?

File: packages/app/src/components/time-series-chart/components/axes/components/axes.tsx [hunk 1]
```diff
@@ -120,58 +100,117 @@ export const Axes = memo(function Axes<T extends TimestampedValue>({
...
-    const preferredDateTicks: number = breakpoints.sm ? (timeframe === 'all' ? (hasDatesAsRange ? 4 : 6) : hasDatesAsRange ? 3 : 5) : hasDatesAsRange ? 2 : 3;
-    const fullDaysInDomain = Math.floor((endUnix - startUnix) / 86400);
-    xTickNumber = Math.max(Math.min(fullDaysInDomain, preferredDateTicks), 2);
-  }
+  const prefferedDateTicksAllTimeFrame = () => {
+    /**
+     * This method gets the difference in years between endUnix and startUnix.
+     * The result is divided by 31557600 and rounded UP with .ceil
+     * 31557600 is calculated based on 86400 (seconds in a day)
+     * multiplied by 365.25 (number of days in a year and .25 to consider leap years)
```
Comment (ben-van-eekelen) [inline packages/app/src/components/time-series-chart/components/axes/components/axes.tsx:108]: Solid description of the 'how', can you also add the 'why' (e.g. the reason why this function is necessary) please? Like you did with the comment on line 145-150.

File: packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import { subtractMonthToDate, extractYearFromDate, formatStyle, getFirstDayOfGivenYear, middleOfDayInSeconds, startOfDayInSeconds } from '@corona-dashboard/common';
+import { Breakpoints } from '~/utils/use-breakpoints';
```
Comment (ben-van-eekelen) [inline packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts:2]: Please order imports alphabetically

File: packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+function getDefault2ValuesForXAxis(startTick: number, endTick: number): TickInstance[] {
+  return [
+    { timestamp: startTick, formatStyle: 'axis-with-day-month-year-short' },
+    { timestamp: endTick, formatStyle: 'axis-with-day-month-year-short' },
+  ] as TickInstance[];
+}
+
+export function createTimeTicksAllTimeFrame(startTick: number, endTick: number, count: number, breakpoints: Breakpoints): TickInstance[] {
```
Comment (ben-van-eekelen) [inline packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts:16]: I don't think `count` is a very descriptive name for the variable holding the amount of ticks to be displayed across the axis. Perhaps you could refactor it to `ticksNumber` or `ticksAmount`?

File: packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import { Breakpoints } from '~/utils/use-breakpoints';
+
+export interface TickInstance {
+  timestamp: number;
+  formatStyle: formatStyle;
+}
+
+function getDefault2ValuesForXAxis(startTick: number, endTick: number): TickInstance[] {
+  return [
+    { timestamp: startTick, formatStyle: 'axis-with-day-month-year-short' },
```
Comment (ben-van-eekelen) [inline packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts:11]: I would suggest to add the constants such as `axis-with-day-month-year-short`, etc. into one of the type definition files in `components/time-series-chart/logic/`. Then you can reference the constants in that file from here and in other components.

File: packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts
Comment (ben-van-eekelen) [inline packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts:1]: Solid work!

File: packages/app/src/utils/use-breakpoints.tsx [hunk 5]
```diff
@@ -3,7 +3,7 @@ import { useTheme } from 'styled-components';
 import { useIsMounted } from './use-is-mounted';
 import { useMediaQuery } from './use-media-query';
 
-type Breakpoints = {
+export type Breakpoints = {
```
Comment (ben-van-eekelen) [inline packages/app/src/utils/use-breakpoints.tsx:6]: Why are you exporting this previously unexported type here? Isn't it better to create a getter function for it like how `rich-content-select.tsx` gets from `use-select.tsx`?

**Review by ben-van-eekelen — DISMISSED** [review]

File: packages/app/src/components/age-demographic/age-demographic-chart.tsx
Comment (VWSCoronaDashboard30) [reply packages/app/src/components/age-demographic/age-demographic-chart.tsx:1]: Good catch!

File: packages/app/src/components/time-series-chart/components/axes/components/axes.tsx (hunk 1 above)
Comment (VWSCoronaDashboard30) [reply packages/app/src/components/time-series-chart/components/axes/components/axes.tsx:108]: Comment enhanced

File: packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts (hunk 4 above)
Comment (VWSCoronaDashboard30) [reply packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts:11]: `axis-with-day-month-year-short` is a `formatStyle` type. I don't need to reference the constant since the type won't allow me to make a typo.

File: packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts (hunk 3 above)
Comment (VWSCoronaDashboard30) [reply packages/app/src/components/time-series-chart/components/axes/logic/create-time-ticks.ts:16]: Changes applied!

File: packages/app/src/utils/use-breakpoints.tsx (hunk 5 above)
Comment (VWSCoronaDashboard30) [reply packages/app/src/utils/use-breakpoints.tsx:6]: The reason this is exported it's because there's logic involved for this in `createTimeTicksAllTimeFrame() function` in a .ts file. Furthermore, `.ts` files cannot use hooks, thus making it impossible to get the breakpoints without a hook.

Comment (ben-van-eekelen) [reply packages/app/src/utils/use-breakpoints.tsx:6]: But wouldn't it then be better to consolidate this type into a seperate `types.ts` or similar file?

**Review by ben-van-eekelen — DISMISSED** [review]

File: packages/app/src/components/time-series-chart/components/axes/components/axes.tsx [hunk 6]
```diff
@@ -120,58 +100,120 @@ export const Axes = memo(function Axes<T extends TimestampedValue>({
...
+
+    return yearsDifferenceInDomain;
+  };
+
+  const bottomAxesTickNumber = useMemo(() => {
+    let value = 2;
+    if (!isPresent(xTickNumber)) {
+      switch (timeframe) {
+        case TimeframeOption.ALL:
+          value = breakpoints.sm ? (hasDatesAsRange ? 4 : prefferedDateTicksAllTimeFrame()) : hasDatesAsRange ? 4 : prefferedDateTicksAllTimeFrame();
```
Comment (netcompany-baam) [inline packages/app/src/components/time-series-chart/components/axes/components/axes.tsx:123]: I find nested ternaries on one line a bit hard to read. I would suggest either changing the formatting to multiple lines and/or using if statements. This style is all over the frontend code as far as I've seen, so it is consistent, but in this case I'd suggest breaking the mold.
